### PR TITLE
chore(repo): setting up the base for migrating to eslint

### PR DESCRIPTION
**Pull request by HuiSF**
#### Description of changes

See each the commit message of each commit.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by ashika112 — APPROVED** [review]
This looks great Hui :pray: Thank you

File: packages/adapter-nextjs/tsconfig.test.json
[the remark this answers is not in the document]
Comment (HuiSF) [reply packages/adapter-nextjs/tsconfig.test.json:1]: These files are now placeholder, they will be used when enabling lint against `__tests__` (they are being ignored in this PR)

File: packages/pubsub/package.json [hunk 1]
```diff
@@ -16,8 +16,7 @@
 		"build-with-test": "npm run clean && npm run build",
 		"build:umd": "webpack && webpack --config ./webpack.config.dev.js",
 		"build:esm-cjs": "rollup -c rollup.config.mjs && cp -R src/vendor dist/cjs/vendor && cp -R src/vendor dist/esm/vendor",
-		"build:cjs:watch": "rimraf dist/cjs && tsc -p tsconfig.watch.json -m commonjs --outDir dist/cjs --watch",
-		"build:esm:watch": "rimraf dist/esm && tsc -p tsconfig.watch.json -m esnext --outDir dist/esm --watch",
+		"build:watch": "cp -R src/vendor dist/cjs/vendor && cp -R src/vendor dist/esm/vendor && rollup -c rollup.config.mjs --watch",
```
[the remark this answers is not in the document]
Comment (HuiSF) [reply packages/pubsub/package.json:19]: Ohh good catch, I thought about it but then forgot

File: .eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+		'.eslintrc.*',
+		'rollup',
+		'rollup.config.*',
+		'setupTests.ts',
+		'jest.setup.*',
+		'jest.config.*',
+		// temporarily disable lint on __tests__
+		'__tests__',
+		// will enable lint by packages
+		// 'adapter-nextjs',
```
[the remark this answers is not in the document]
Comment (HuiSF) [reply .eslintrc.js:41]: Yes, exactly.

File: package.json [hunk 3]
```diff
@@ -84,11 +85,24 @@
...
 		"@size-limit/webpack-why": "^8.1.0",
+		"@stylistic/eslint-plugin": "^1.5.4",
 		"@types/jest": "^29.5.8",
 		"@types/lodash": "4.14.182",
 		"@types/node": "^8.9.5",
 		"@types/puppeteer": "1.3.0",
+		"@typescript-eslint/eslint-plugin": "^6.20.0",
+		"@typescript-eslint/parser": "^6.20.0",
 		"babel-loader": "^8.3.0",
+		"eslint": "^8.56.0",
```
[the remark this answers is not in the document]
Comment (HuiSF) [reply package.json:96]: You meant the tslint related packages? If so, yes, we can remove those once migration is completed.

File: packages/pubsub/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (HuiSF) [reply packages/pubsub/package.json]: Added `mkdirp` in the workspace, and create the `vendor` directories for pubsub while enabling the watch mode.

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (jimblanc) [reply package.json:96]: Sorry yeah, tslint related deps

**Review by jimblanc — APPROVED** [review]

File: .vscode/settings.json [hunk 4]
```diff
@@ -1,12 +1,13 @@
...
 	"editor.tabSize": 4,
 	"prettier.requireConfig": true,
-	"typescript.tsdk": "node_modules/typescript/lib"
+	"typescript.tsdk": "node_modules/typescript/lib",
+	"formattingToggle.affects": [
+		"editor.codeActionsOnSave.source.fixAll.eslint",
+		"editor.formatOnPaste",
+		"editor.formatOnSave",
+		"editor.formatOnType"
+	]
```
Comment (calebpollman) [inline .vscode/settings.json:12]: This feels a bit restrictive/opinionated, what is the context of this addition?

File: .eslintrc.js [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+			'error',
+			{
+				vars: 'all',
+				varsIgnorePattern: '^_',
+				args: 'after-used',
+				argsIgnorePattern: '^_',
+			},
+		],
+		'valid-typeof': ['error', { requireStringLiterals: false }],
+		'@stylistic/comma-dangle': [
```
Comment (calebpollman) [inline .eslintrc.js:126]: This reads as something that Prettier could just handle?

Comment (HuiSF) [reply .eslintrc.js:126]: Prettier handles this (when it actually reformats), this rule is for showing errors with the eslint in details.

File: .vscode/settings.json (hunk 4 above)
Comment (HuiSF) [reply .vscode/settings.json:12]: This is not restrictive, IMO, but gives the most freedom.
* reformat on save is not enabled by default
* ppl who are using vscode can install the recommended extension formatting toggle to enable reformat on save (this is the existing set up prior to this migration)
* for ppl who are not using auto reformat at all, they will catch formatting issues when run `yarn test`

**Review by jimblanc — APPROVED** [review]

File: .eslintrc.js [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	},
+	extends: [
+		'eslint:recommended',
+		'standard',
+		'plugin:import/errors',
+		'plugin:import/recommended',
+		'plugin:import/typescript',
+		'plugin:@typescript-eslint/stylistic',
+		'plugin:@typescript-eslint/recommended',
+		"plugin:prettier/recommended",
```
Comment (calebpollman) [inline .eslintrc.js:16]: Would be careful with usage of `"plugin:prettier/recommended"`. Mixing linting with formatting can lead to undesired conflicts

File: .vscode/settings.json (hunk 4 above)
Comment (calebpollman) [reply .vscode/settings.json:12]: Got it, misread these as enforced

File: .eslintrc.js (hunk 5 above)
Comment (calebpollman) [reply .eslintrc.js:126]: Feels extraneous IMO but do not have a strong opinion

File: .eslintrc.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+		project: './tsconfig.json',
+	},
+	extends: [
+		'eslint:recommended',
+		'standard',
+		'plugin:import/errors',
+		'plugin:import/recommended',
+		'plugin:import/typescript',
+		'plugin:@typescript-eslint/stylistic',
+		'plugin:@typescript-eslint/recommended',
```
Comment (calebpollman) [inline .eslintrc.js:15]: Is there any particular reasoning for usage of the `recommended` rule set over `recommended-type-checked`?

File: .eslintrc.js [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+		],
+		'import/no-deprecated': 'warn',
+		'import/no-empty-named-blocks': 'error',
+		'import/no-mutable-exports': 'error',
+		'import/no-relative-packages': 'error',
+		'import/newline-after-import': 'error',
+		'import/order': ['error', { 'newlines-between': 'always' }],
+		'no-eval': 'error',
+		'no-param-reassign': 'error',
+		'no-shadow': 'off',
```
Comment (calebpollman) [inline .eslintrc.js:102]: Definitely an opinion, but turning off `no-shadow` can lead to code that is difficult to grok as a reader

Comment (calebpollman) [reply .eslintrc.js:102]: Just saw that this is turned off in favor of the `@typescript-eslint` version 😅 

For future maintainers, might be helpful to note which rules are superseded by corresponding `@typescript-eslint` rules

**Review by calebpollman — APPROVED** [review]
Left some thoughts/questions mostly from my own battles with configuring `eslint`. 

Definitely an improvement overall, and also plan on using some of these in the UI repo 😏

File: .eslintrc.js (hunk 7 above)
Comment (HuiSF) [reply .eslintrc.js:15]: Unfortunately that requires a much heavier code migration on some packages which we don't have enough bandwidth to do so :/ would like to use this as a starter to get the codebase in a better place, and then apply more restrict linter rules.

**Review by iartemiev — APPROVED** [review]

**Pull request merged**